core/src/main/java/org/opensearch/sql/executor/QueryService.java

```diff
@@ -132,7 +137,31 @@ public void executePlanByCalcite(
       RelNode plan,
       CalcitePlanContext context,
       ResponseListener<ExecutionEngine.QueryResponse> listener) {
-    executionEngine.execute(optimize(plan), context, listener);
+    executionEngine.execute(convertToCalcitePlan(optimize(plan)), context, listener);
+  }
+
+  /**
+   * Convert OpenSearch Plan to Calcite Plan. Although both plans consist of Calcite RelNodes, there
+   * are some differences in the topological structures or semantics between them.
+   *
+   * @param osPlan Logical Plan derived from OpenSearch PPL
+   */
+  private static @NotNull RelNode convertToCalcitePlan(RelNode osPlan) {
+    RelNode calcitePlan = osPlan;
+
+    /* Calcite only ensures collation of the final result produced from the root sort operator.
+     * While we expect that the collation can be preserved through the pipes over PPL, we need to
+     * explicitly add a sort operator on top of the original plan
+     * to ensure the correct collation of the final result.
+     * See logic in ${@link CalcitePrepareImpl}
+     * For the redundant sort, we rely on Calcite optimizer to eliminate
+     */
+    RelCollation collation = osPlan.getTraitSet().getCollation();
+    if (!(osPlan instanceof Sort) && collation != RelCollations.EMPTY) {
+      calcitePlan = LogicalSort.create(osPlan, collation, null, null);
```

Comment:
Any reason for adding this NotNull annotation? Would be better to add this?
```
if (calcitePlan = null) return osPlan;
```

Reply:
> While we expect that the collation can be preserved through the pipes over PPL

could u elberate more on this? i do not think PPL command should preserved order.

Reply:
@qianheng-aws @LantaoJin 
I seem, some PPL command should preserved order, for instance, `fields`, `take`, but others not, `stats`.
In ANSI SQL, `select a from tbl order by b`, order is perserved becuase of logically processing order of select statement is acutall `from -> select -> order`.